activities.el

```diff
@@ -801,6 +851,116 @@ activity's name is NAME."
+  (lambda (name)
+    (when-let ((activity (map-elt activities-activities name)))
+      (let (data)
+	(dolist (type '(last default))
+	  (when-let ((state (cl-struct-slot-value 'activities-activity type activity)))
+	    (let* ((time (map-elt (activities-activity-state-etc state) 'time))
+		   (buffers-and-files (activities--buffers-and-files state)))
+	      (setf (alist-get type data)
+		    (list (and time (float-time (time-since time))) buffers-and-files)))))
+	(pcase-let* ((`(,default-age ,default-buffers-and-files) (map-elt data 'default))
+		     (`(,last-age ,last-buffers-and-files) (map-elt data 'last)) ;possibly nil
+		     (age (if last-age (min last-age default-age) default-age))
+		     (buffers-and-files (if last-age
+					    last-buffers-and-files
+					  default-buffers-and-files))
+		     (num-buffers (length buffers-and-files))
+		     (num-files (seq-count #'stringp (mapcar #'cdr buffers-and-files)))
+		     (dirtyp (when last-buffers-and-files
+			       (activities--buffers-and-files-differ-p
+				last-buffers-and-files
+				default-buffers-and-files)))
+		     (annotation (format "%s%s buf%s %s file%s "
+					 (if (activities-activity-active-p activity)
+					     (propertize "@" 'face 'bold) " ")
+					 (propertize (format "%2d" num-buffers) 'face 'success)
+					 (if (= num-buffers 1) " " "s")
+					 (propertize (format "%2d" num-files) 'face 'warning)
+					 (if (= num-files 1) " " "s")))
+		     (age-color (or (cdr (vc-annotate-compcar
+					  (* (/ age max-age) oldest-possible)
+					  vc-annotate-color-map))
+				    vc-annotate-very-old-color))
+		     (age-annotation (propertize
+				      (format "%15s" (apply #'format "[%d %s]"
+							    (activities--age age)))
+				      'face `(:foreground ,age-color
+							  :background ,vc-annotate-background)))
+		     (dirty-annotation (if dirtyp (propertize "*" 'face 'bold) " ")))
+	  (concat (propertize " " 'display
+			      `(space :align-to (- right ,(+ 1 (length annotation)
+							     (length age-annotation)))))
+		  annotation age-annotation dirty-annotation))))))
+
+(defun activities-sort-by-active-age (names)
+  "Return the list activity NAMES sorted active first, then by age."
+  (sort names
+	(lambda (a b)
+	  (let* ((activity-a (map-elt activities-activities a))
+		 (state-a (or (activities-activity-last activity-a)
+			      (activities-activity-default activity-a)))
+		 (time-a (map-elt (activities-activity-state-etc state-a) 'time))
+		 (activep-a (activities-activity-active-p activity-a))
+		 (activity-b (map-elt activities-activities b))
+		 (state-b (or (activities-activity-last activity-b)
+			      (activities-activity-default activity-b)))
+		 (time-b (map-elt (activities-activity-state-etc state-b) 'time))
+		 (activep-b (activities-activity-active-p activity-b)))
+	    (or (and activep-a (not activep-b)) (time-less-p time-b time-a))))))
+
+(defun activities--completion-table (activities)
```

Comment:
Do you foresee this function being called elsewhere than in `activities-completing-read`?  If not, I think I'd prefer for it to be defined inside that function, to avoid the top-level symbol.

Reply:
Done.